ttnn/cpp/ttnn/operations/matmul/device/matmul_op.hpp

```diff
@@ -173,7 +173,8 @@ struct Matmul {
 
     void validate(
         const std::vector<Tensor>& input_tensors,
-        const std::vector<std::optional<const Tensor>>& optional_input_tensors) const;
+        const std::vector<std::optional<const Tensor>>& optional_input_tensors,
+        const std::vector<std::optional<Tensor>>& optional_output_tensors = {std::nullopt}) const;
     std::vector<ttnn::SimpleShape> compute_output_shapes(const std::vector<Tensor>& input_tensors) const;
```

Comment:
Take a look https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operation.hpp#L303
You need to implement compute_output_specs in general and then the code should allow you to use optional output tensors.